### ISSUE TITLE
Handle external URL references in archives to enable offline use in Kolibri

### ISSUE THREAD

**Issue by ivanistheone** [learningequality/ricecooker#233]
### Description

Archive-based content types (H5P, HTML5 zip, IMSCP) may contain references to external URLs — images, videos, fonts, stylesheets, scripts — that won't be available in offline Kolibri deployments. The pipeline conversion handlers need to scan archive contents for these references, download the resources, bundle them into the archive, and rewrite the references to point to the local copies.

#### Existing logic in `downloader.py`

We already have most of this URL extraction and rewriting logic in `downloader.py`. `download_static_assets()` and its inner functions extract URLs from:
- HTML attributes: `img[src]`, `link[href]`, `script[src]`, `source[src]`, `img[srcset]`, `[style*="background-image"]`
- CSS: `url()` references for fonts, images, etc.
- Recursive resource following (CSS that references fonts, etc.)

And `ArchiveDownloader` downloads pages with all their resources, rewrites paths, and creates ZIP archives.

However, this logic is **not reusable by the pipeline** because:
- It's tightly coupled to web scraping (takes URLs, not archive contents)
- The extraction/rewriting logic is buried in inner functions that can't be unit tested (see #303)
- It has known bugs being fixed in #636 and #639

#### kolibri-zip as reference spec

Kolibri's [`kolibri-zip`](https://github.com/learningequality/kolibri/blob/release-v0.19.x/packages/kolibri-zip/src/index.js) package handles the *runtime* side of this: when rendering ZIP-based content, it extracts files, resolves internal path references, and rewrites them to blob URLs. Its [`fileUtils.js`](https://github.com/learningequality/kolibri/blob/release-v0.19.x/packages/kolibri-zip/src/fileUtils.js) provides a comprehensive spec for which reference types need handling:

**HTML/XML files** (`src`, `href`, `srcset`, inline `style`, `<style>` blocks):
```html
<img src="images/photo.jpg">
<link href="styles/main.css">
<script src="https://cdn.example.com/lib.js"></script>
<img srcset="img-300.jpg 300w, img-600.jpg 600w">
<div style="background: url('bg.png')">
```

**CSS files** (`url()`, `@import` — both `url()` and bare string forms):
```css
@import 'fonts/custom.css';
background-image: url('../images/bg.png');
@font-face { src: url('https://fonts.example.com/font.woff2'); }
```

**H5P JSON** (`path` attributes in `content/content.json`):
```json
{
  "video": {
    "files": [
      { "path": "https://h5p.org/sites/default/files/h5p/iv.mp4", "mime": "video/mp4" }
    ]
  }
}
```

kolibri-zip handles *internal* references at runtime, but cannot fetch *external* URLs (especially offline). That's ricecooker's job at import time.

### Key architectural constraint

External resource downloading and reference rewriting must happen **before** `create_predictable_zip` is called in `ArchiveProcessingBaseHandler.handle_file()` ([convert.py](https://github.com/learningequality/ricecooker/blob/develop/ricecooker/utils/pipeline/convert.py)). Once `create_predictable_zip` runs, the archive is sealed — it iterates existing files and can only transform them via `file_converter` (currently used for media compression), not add new ones.

The flow in `handle_file()` would become:

```
validate_archive(path)
path = download_and_rewrite_external_refs(path)   # <-- new step
create_predictable_zip(path, file_converter=...)   # existing step
```

Where `download_and_rewrite_external_refs` would:
1. Extract the archive to a temp directory
2. Scan text-based files for external URL references
3. Download those resources into the temp directory
4. Rewrite references in the source files to point to local copies
5. Return the temp directory path (which `create_predictable_zip` already accepts — it handles both directories and zip files)

### Related issues and PRs

- **#303** — Refactor `downloader.py` to make URL detection and rewriting unit-testable. **This issue supersedes #303** — the extraction of URL logic into shared utilities serves both the testability goal and the pipeline integration goal.
- **#636** — Fix CSS `link` tag filtering (`"rel" in node` vs `"rel" in node.attrs`) and extensionless URL path placement. By @jaltekruse.
- **#639** — Fix CSS `@import` with bare strings (not wrapped in `url()`). By @jaltekruse.

#### Phase 1: Land bug fixes from #636 and #639

Merge the open PRs first to preserve @jaltekruse's contributor attribution in git history before the refactor changes the code structure. These fix real bugs in the URL extraction logic that the shared utilities will need to carry forward.

#### Phase 2: Extract shared utilities from `downloader.py` (supersedes #303)

Extract the URL extraction and rewriting logic from `download_static_assets()` inner functions into standalone, testable utility functions that operate on file contents (strings/bytes) rather than requiring a live HTTP session:

- **URL extraction**: Given HTML/CSS/JSON content, return a list of referenced URLs
- **URL rewriting**: Given content and a URL mapping (old → new), return rewritten content
- **External URL filtering**: Distinguish external (http/https) from internal (relative paths already in archive) references

This directly addresses #303's concern about untestable inner functions. The extracted functions can be unit tested with plain strings — no HTTP server, no filesystem, no platform-specific path issues. This also resolves the Windows test failures in #636, since the core logic tests won't depend on filesystem paths.

#### Phase 3: Create archive processing utility

Build on the Phase 2 utilities to create an archive-level processor:
- Open an archive and iterate its text-based files (HTML, CSS, JSON, XML)
- Use Phase 2 extractors to find external URL references
- Download external resources into the extracted archive directory
- Use Phase 2 rewriters to update references to local paths
- Loop detection for recursive references (à la kolibri-zip's `visitedPaths`)

#### Phase 4: Integrate into pipeline conversion handlers

Wire the archive processor into the existing handlers, running before `create_predictable_zip`:
- **`H5PConversionHandler`**: Scan `content/content.json` for external `path` values, plus HTML/CSS in content (highest priority — videos and images commonly external)
- **`HTML5ConversionHandler`**: Scan HTML/CSS files for external references
- **`IMSCPConversionHandler`**: Scan entry point HTML files and their CSS for external references

### Reference types to handle

From kolibri-zip's `fileUtils.js` and existing `downloader.py` logic:

| File type | Reference patterns | Source |
|---|---|---|
| HTML/XML | `src`, `href`, `srcset` attributes; inline `style`; `<style>` blocks | kolibri-zip `DOMMapper`, downloader.py `download_static_assets()` |
| CSS | `url()`, `@import` (both `url()` and bare string forms) | kolibri-zip `CSSMapper`, downloader.py `_CSS_URL_RE`, PR #639 `_CSS_IMPORT_RE` |
| H5P JSON | `path` attributes in `content/content.json` | H5P-specific |

### References

- kolibri-zip source: https://github.com/learningequality/kolibri/tree/release-v0.19.x/packages/kolibri-zip/src
- Current pipeline handlers: [`ricecooker/utils/pipeline/convert.py`](https://github.com/learningequality/ricecooker/blob/develop/ricecooker/utils/pipeline/convert.py)
- Existing download logic: [`ricecooker/utils/downloader.py`](https://github.com/learningequality/ricecooker/blob/develop/ricecooker/utils/downloader.py)

**Comment by ivanistheone** [comment]
Here are some h5p content kinds that we'll need "adapers" for:
  - Image assets in timeline (see [example here](http://35.185.120.102/en/learn/#/topics/c/9f0297b9648d542698f048490b946e77))
  - [iFrame embedder](http://35.185.120.102/en/learn/#/topics/c/b135832b388b5c22a2efa10da2fece3e)
  - video assets in [Image Hotspots](http://35.185.120.102/en/learn/#/topics/c/84703961e747568696509b98e8bdd68b) (click on the biggest strawberry)
  - (minor) Web font resources might need to offlined (see [here](http://35.185.120.102/en/learn/#/topics/c/70e9dcd73c6d504393b8ce54297b3d36))

We'll have to think about something generic that can be used by ricecooker and by Studio (when users upload through the Studio web interface).



Here are some content types which we cannot hope to get working:
  - [Audio Recorder](http://35.185.120.102/en/learn/#/topics/c/b3c2816091a1535da9f7df56b74765d0)
  - [Speak the Words](http://35.185.120.102/en/learn/#/topics/c/a5e38c00aeeb514990f0fef6ff0d6c0c) and [Speak the Words Set](http://35.185.120.102/en/learn/#/topics/c/c718b9e173dc50beb4764ce07c8b14e3). Does not load at all in FF, and doesn't seem to work in Chrome.

We should probably remove these from the new demo channel.

**Comment by intelliant01** [comment]
@ivanistheone Regarding -

> Here are some content types which we cannot hope to get working:
>  * [Audio Recorder](http://35.185.120.102/en/learn/#/topics/c/b3c2816091a1535da9f7df56b74765d0)
>   * [Speak the Words](http://35.185.120.102/en/learn/#/topics/c/a5e38c00aeeb514990f0fef6ff0d6c0c) and [Speak the Words Set](http://35.185.120.102/en/learn/#/topics/c/c718b9e173dc50beb4764ce07c8b14e3). Does not load at all in FF, and doesn't seem to work in Chrome.

Have you tried on a server with https and Chrome? For example through kolibridemo?

**Comment by rtibblesbot** [comment]
## Phase 2 Complete: URL Utilities

Implemented `ricecooker/utils/url_utils.py` with 8 pure-function utilities for detecting and rewriting external URL references in HTML, CSS, and H5P JSON content:

- `is_external_url()`, `derive_local_filename()`
- `extract_urls_from_html()`, `extract_urls_from_css()`, `extract_urls_from_h5p_json()`
- `rewrite_urls_in_html()`, `rewrite_urls_in_css()`, `rewrite_urls_in_h5p_json()`

Incorporates bug fixes from PRs #636 (`css_node_filter` attr check) and #639 (CSS `@import` bare string regex).

64 new unit tests passing — all pure-function tests, no I/O.

**Phase 1 note**: PRs #636 and #639 are still open. The fixes have been incorporated directly into the new utilities. Those PRs can still be merged independently for the existing `downloader.py` code.

Moving to Phase 3: Archive external reference processor.

**Comment by jaltekruse** [comment]
Hey @rtibbles , I saw that the agent you tried assigning to this issue mentioned my PRs. I'd be open to taking a crack at doing this refactoring. If there is any additional context you would like to share beyond what is documented here, and what I can read in your conversation with the LLM on its attempt, happy to get in the best position for success before starting this work.

Also as a small side-note I updated this PR and I think it should be in a decent state, but I'd rather not get the dev environment set up in native windows if I can avoid it for now, so if you could approve the CI run of the tests I can see if I did actually fix the test. https://github.com/learningequality/ricecooker/pull/636

**Comment by rtibblesbot** [comment]
## Triage: Prerequisites blocking implementation

This issue is well-specified with a clear phased approach. However, **Phase 1 explicitly requires merging PRs #636 and #639 before the refactor can begin**, and both are currently open:

- **PR #636** (Fix CSS link tag filtering + extensionless URL paths) — Open, needs review/merge
- **PR #639** (Fix CSS @import bare strings) — Open, WIP, needs review/merge

These need to land first to preserve @jaltekruse's contributor attribution before the refactor changes the code structure around these fixes.

### Stakeholder comment acknowledgment

- **@ivanistheone's** list of H5P content types needing adapters (timeline images, iframe embedder, video in Image Hotspots, web fonts) is consistent with the issue's Phase 3/4 scope for H5P JSON path extraction and HTML/CSS reference handling. These specific content patterns will be covered by the generic external URL extraction approach described in the issue.
- **@jaltekruse** has expressed interest in doing this refactoring work — coordination on assignment may be needed before an agent proceeds.
- **@intelliant01's** note about HTTPS requirements for Audio Recorder/Speak the Words is relevant context but outside the scope of this issue (those are runtime browser API limitations, not an archive processing concern).

**Blocked on:** PRs #636 and #639 being reviewed and merged.

**Comment by rtibblesbot** [comment]
## Triage Assessment

This issue is well-specified with clear requirements, a phased approach, and detailed reference types to handle.

### Blocking Prerequisites

**Phase 1 requires merging PRs #636 and #639 before implementation can begin:**

- **PR #636** ("Fix downloading a linked css resource with no extension") by @jaltekruse — currently OPEN and MERGEABLE. Needs CI approval and maintainer review.
- **PR #639** ("Fix rewriting of CSS @import statements that use strings instead of url()") by @jaltekruse — currently OPEN and MERGEABLE. Needs maintainer review.

The issue explicitly states these should be merged first to preserve contributor attribution before the refactor changes the code structure. Implementation of Phases 2-4 cannot proceed until these are landed.

### Stakeholder Comment Acknowledgment

- **@ivanistheone's** examples of H5P content types (Timeline images, iFrame embedder, Image Hotspots videos, web fonts) confirm the need for a generic external resource downloading approach. These specific cases will inform the H5P JSON path extraction in Phase 3/4 and the note about making the solution reusable by Studio aligns with extracting standalone utilities in Phase 2.
- **@intelliant01's** note about HTTPS requirements for Audio Recorder / Speak the Words is useful context but doesn't affect the implementation approach for resource downloading.
- **@jaltekruse** has expressed interest in contributing to this refactoring and has the open prerequisite PRs. Coordination with them on Phase 2 (which restructures the code their PRs touch) would be valuable once the PRs are merged.

### Next Steps

Waiting for PRs #636 and #639 to be reviewed and merged before proceeding with implementation planning for Phases 2-4.

**Comment by rtibblesbot** [comment]
This issue has blocking prerequisites: PRs #636 and #639 need to be merged before implementation can begin (Phase 1). Both are currently still open. The issue explicitly requires these land first to preserve contributor attribution and carry forward bug fixes into the shared utilities.

Ready to begin Phase 2+ once those PRs are merged.

**Comment by rtibblesbot** [comment]
**Plan:** Pipeline conversion handlers for archive formats (HTML5, H5P) scan their contents for external URL references, download those resources into the archive, and rewrite references to the local copies — so the content works offline in Kolibri.

- [ ] External-URL filter + CSS extraction/rewriting
- [ ] HTML extraction/rewriting
- [ ] H5P JSON `path` extraction/rewriting
- [ ] Point `downloader.py` at the shared regexes (dedupe, no behavior change)
- [ ] `download_and_rewrite_external_refs`
- [ ] Opt-in external-ref download in `ArchiveProcessingBaseHandler`

---
_@rtibblesbot's comments are generated by an LLM, and should be evaluated accordingly_

<details><summary>How was this generated?</summary>

- Implemented code following the pre-approved plan using test-driven development (red-green-refactor)
- Ran the project's full test suite and linter before committing
- For UI changes: captured live screenshots to verify rendering
- For backend changes: verified with example API/CLI output

</details>

**Comment by rtibbles** [comment]
## Revised architecture — supersedes the plan above; PR #690 to be reworked against this

Grounded in the actual runtime (kolibri-zip `fileUtils.js`, kolibri-sandbox `H5PRunner.js`) and ricecooker's `FilePipeline`.

### Decisions

- **One detector.** `references.py` is a faithful Python port of kolibri-zip's mappers (CSS `url()`/`@import`; DOM `src`/`href`/`srcset`/inline + `<style>`), so build-time and runtime detection agree. #690 wrote a second detector alongside `download_static_assets`, leaving #303 unresolved.
- **Delete `downloader.py`.** The pipeline is the sole download-and-rewrite path; no backwards-compat shim.
- Its scraping code has no production callers; only `read` (html_writer, pdf) and `get_archive_filename`/thumbnail (chefs) are live, and migrate to the pipeline.
- **Re-enter the *same* pipeline per referenced file**, via `Handler.parent` — not the module-global `ConversionStageHandler` #690 constructed.
- Each ref goes DOWNLOAD → CONVERT (fetch external / read internal; compress media; convert incompatible images), then is rewritten to the processed output in lockstep.
- That walk replaces the media-only `_read_and_compress_archive_file`.
- **No pruning.** The runtime resolves paths dynamically and serves on demand, so static reachability under-approximates; keep every file, only add downloaded externals.
- **Surgical HTML rewriting.** Splice only matched spans; #690's full-DOM `str(soup)` round-trip corrupts third-party HTML5 apps.
- **Default placement:** downloaded asset adjacent to its referencing file, relative rewrite.
- **H5P placement:** `content.json` externals go into `content/` with a content-relative path so `H5P.getPath` resolves them; #690's `../_static/` rewrite will not.
- **All archive formats process HTML/CSS refs**, including EPUB/KPUB/Bloom.
- **Do not hardcode `index.html`.** Ref processing walks all files, staying compatible with #672's denesting and non-root entry points.

### Build plan

- [ ] `references.py` mappers (kolibri-zip port): CSS + HTML/DOM, surgical rewrite, `html5lib`-located spans
- [ ] Reference-driven archive processing: walk files, re-enter the same `FilePipeline` per ref, visited-guard recursion, per-format placement
- [ ] Wire into `ArchiveProcessingBaseHandler`; remove `_read_and_compress_archive_file` and the `_get_conversion_stage` singleton
- [ ] H5P `H5PContentMapper` (content.json `path`s) → download into `content/`, content-relative rewrite
- [ ] Delete `downloader.py`; migrate `read` (html_writer, pdf) and `get_archive_filename`/thumbnail (chefs) to the pipeline
- [ ] Rewrite `tests/test_downloader.py` as `references.py` + pipeline tests
- [ ] Follow-up issue: headless JS-render DOWNLOAD handler (dropped here)

**Comment by rtibblesbot** [comment]
The bot has been waiting for a response for 2 hour(s). It will resume automatically when someone replies.

**Comment by rtibblesbot** [comment]
The bot has been waiting for a response for 2 hour(s). It will resume automatically when someone replies.

**Comment by rtibblesbot** [comment]
The bot has been waiting for a response for 2 hour(s). It will resume automatically when someone replies.

**Comment by rtibblesbot** [comment]
The bot has been waiting for a response for 2 hour(s). It will resume automatically when someone replies.
